doc/dev/crimson/poseidonstore.rst

```diff
@@ -0,0 +1,401 @@
+  +-----------------------------------------------------------------------------+
+  | WAL -> |      | Super block | Object meta | Allocation bitmap | Data blocks |
+  +-----------------------------------------------------------------------------+
+  | Sharded partition N 
+  +-----------------------------------------------------------------------------+
+  | WAL -> |      | Super block | Object meta | Allocation bitmap | Data blocks |
+  +-----------------------------------------------------------------------------+
+
+
+- WAL
+  - Log and frequently updated metadata are stored as a WAL entry in the WAL partition
+  (Can be placed on NVM)
+  - Space within the WAL partition is continually reused in a circular manner
+  - Flush the metadata if necessary
+- Write procedure for Metadata
+  - Appended at the WAL first
+  - Overwrite the metadta in the data partition when flushing
+- Write procedure for Data
+  - Overwrite the data in the data partition
+- Disk layout
+  - Object meta can embed data. For example, object_info_t can be recorded as an entry of
+    the Object meta
+  - Allocation bitmap manages the Data blocks
+  - Super block manages data partitions
+
+
+I/O procedure
+-------------
+- Write
+  For incoming writes, data is handled differently depending on the request size; 
+  data is either written twice (WAL) or written in a log-structured manner.
+
+  (1) If Request Size ≤ Threshold (similar to minimum allocation size in BlueStore)
+    Write data and metadata to [WAL] —flush—> Write them to [Data section (in-place)] and 
+    [Metadata section], respectively.
+
+    Since the CPU becomes the bottleneck for small I/O workloads, in-place update scheme is used.
+    Double write for small data may be better than host-GC overhead in terms of CPU consumption 
+    in the long run
+
+  (2) Else if Request Size > Threshold
+    Append data to [Data section (log-structure)] —> Write the corresponding metadata to [WAL] 
+    —flush—> Write the metadata to [Metadata section]
+
+    For large I/O workloads, the host can afford process host-GC
+    Also, the host can garbage collect invalid objects more effectively when their size is large
+
+    Note that Threshold can be configured to a very large number so that only the scenario (1) occurs.
+    With this design, we can control the overall I/O procedure with the optimizations for crimson
+    as described above.
+
+  - Detailed flow
+    - 1. Append a log entry that contains pg_log, snapset, object_infot_t, block allocation
+      using NVMe atomic write command on the WAL
+      - NVMe provides atomicity guarantees for a write command (Atomic Write Unit Power Fail)
+        For example, 512 Kbytes of data can be atomically written at once without fsync()
+      - Small size (object_info_t, snapset, collection, etc.) can be embed
+      - stage 1
+        WAL (written) --> | TxBegin A | Log Entry | TxEnd A | (if small)
+        Data partition (written)--> | Data blocks | (if large)
```

Comment:
I don't understand what this means.  For a large write, I'd assume that you need to write and persist the data portion *prior* to the WAL metadata update (as bluestore does).

Reply:
Wait, is the idea that for a large write you'd do *three* writes?  A WAL (tx begin) ->data write->WAL (tx end)?